addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-upload-file-watcher.go

```diff
@@ -2,132 +2,295 @@
+					log.Println("Unable to stat file", err)
+					return
+				}
+
+				if fi.Mode().IsDir() {
+					if event.Op&fsnotify.Create == fsnotify.Create {
+						if debug {
+							log.Println(event.Name, "is a new directory, watching")
+						}
+						if err := filepath.Walk(
+							event.Name,
+							watchDirFactory(watcher, dirRequiringCrawlCh, debug),
+						); err != nil {
+							log.Println("Unable to watch directory", err)
+						}
+						return
 					}
-					// First check that this is a file
-					fi, err := os.Stat(event.Name)
-					if err != nil {
-						log.Println("Unable to stat file", err)
-						continue
+					if debug {
+						log.Println(event.Name, "is a directory, skipping")
 					}
+					return
+				}
 
-					if fi.Mode().IsDir() {
-						if event.Op&fsnotify.Create == fsnotify.Create {
-							if debug {
-								log.Println(event.Name, "is a new directory, watching")
-							}
-							if err := filepath.Walk(
-								filepath.Join(syncdir, event.Name),
-								watchDirFactory(watcher, debug),
-							); err != nil {
-								log.Println("Unable to watch directory", err)
-							}
-							continue
-						}
+				uploadToS3(sess, syncDir, event.Name, bucket, prefix, kmsKeyId, debug)
+			}
+		}
+
+		uploadDir := func(dirToUpload string, debug bool) {
+			if debug {
+				log.Println("Crawling directory", dirToUpload, "to upload file to s3 who may have been missed by file watcher")
+			}
+			if err := filepath.Walk(
+				dirToUpload,
+				func(path string, fi os.FileInfo, err error) error {
+					if fi != nil && !fi.Mode().IsDir() {
 						if debug {
-							log.Println(event.Name, "is a directory, skipping")
+							log.Println("Uploading file", path, "to S3")
 						}
-						continue
+						uploadToS3(sess, syncDir, path, bucket, prefix, kmsKeyId, debug)
+						return nil
 					}
+					return nil
+				},
+			); err != nil {
+				log.Println("Unable to upload directory", err)
+			}
+		}
 
-					uploadToS3(sess, syncdir, event.Name, bucket, prefix, kmsKeyId, debug)
-				}
+		timeOut := func() {
+			continueFileUploadWatcher = false
+
+			// Decrement from the wait group indicating we are done
+			wg.Done()
+
+			// Close the watcher once we are done watching file changes
+			watcher.Close()
+		}
 
-			case err := <-watcher.Errors:
-				log.Println("error:", err)
+		for continueFileUploadWatcher {
+			if stopUploadWatchersAfter > 0 {
+				select {
+				case <-time.After(time.Duration(stopUploadWatchersAfter) * time.Second):
+					timeOut()
+					break
+				case dirToUpload := <-dirRequiringCrawlCh:
+					uploadDir(dirToUpload, debug)
+				case event := <-watcher.Events:
+					processFileWatcherEvent(&event)
+				case err := <-watcher.Errors:
+					log.Println("error:", err)
+				}
+			} else {
+				select {
+				case event := <-watcher.Events:
+					processFileWatcherEvent(&event)
+				case dirToUpload := <-dirRequiringCrawlCh:
+					uploadDir(dirToUpload, debug)
+				case err := <-watcher.Errors:
+					log.Println("error:", err)
```

Comment:
Can this duplicate code be moved to a method so that it can be referenced in stopUploadWatchersAfter and non stopUploadWatchersAfter cases.

Reply:
Not that I know of. They are two different `select` statements one with a timeout and other one without timeout. These `select` statements are different from the usual `switch` statements we see in other languages. The `select` here is waiting to receive events from 3 channels vs 4 channels (one channel being the one for timeout). To minimize duplication I had already moved all case handling logic to separate methods the only duplication at the moment is two `select` statements receiving from 3 vs 4 channels.